BraveWallet/Extensions/RpcServiceExtensions.swift

```diff
@@ -83,42 +117,109 @@ extension BraveWalletJsonRpcService {
+            ) {
+              completion(BDouble(valueString))
+            } else {
+              completion(nil)
+            }
+          }
         )
-      } else if token.isErc721 {
-        erc721TokenBalance(
-          token.contractAddress,
-          tokenId: token.tokenId,
-          accountAddress: accountAddress,
-          chainId: network.chainId,
-          completion: convert)
-      } else {
+      case .sol:
+        if token.symbol == network.nativeToken.symbol {
+          solanaBalance(accountAddress, chainId: network.chainId) { lamports, status, errorMessage in
+            guard status == .success else {
+              completion(nil)
+              return
+            }
+            let formatter = WeiFormatter(decimalFormatStyle: decimalFormatStyle)
+            if let valueString = formatter.decimalString(
+              for: "\(lamports)",
+              radix: .decimal,
+              decimals: Int(token.decimals)
+            ) {
+              completion(BDouble(valueString))
+            } else {
+              completion(nil)
+            }
+          }
+        } else {
+          splTokenAccountBalance(
+            accountAddress,
+            tokenMintAddress: token.contractAddress,
+            chainId: network.chainId
+          ) { amount, lamports, uiAmountString, status, errorMessage in
+            guard status == .success else {
+              completion(nil)
+              return
+            }
+            let formatter = WeiFormatter(decimalFormatStyle: decimalFormatStyle)
+            if let valueString = formatter.decimalString(
+              for: "\(amount)",
+              radix: .decimal,
+              decimals: Int(token.decimals)
+            ) {
+              completion(BDouble(valueString))
+            } else {
+              completion(nil)
+            }
+          }
+        }
+      case .fil:
+        completion(nil)
+      @unknown default:
         completion(nil)
       }
     }
   }
+  
+  private func ethBalance(
+    for token: BraveWallet.BlockchainToken,
+    in accountAddress: String,
+    network: BraveWallet.NetworkInfo,
+    completion: @escaping (String, BraveWallet.ProviderError, String) -> Void
+  ) {
+    if token.symbol == network.symbol {
+      balance(
+        accountAddress,
+        coin: .eth,
+        chainId: network.chainId,
+        completion: completion
+      )
+    } else if token.isErc20 {
+      erc20TokenBalance(
+        token.contractAddress(in: network),
+        address: accountAddress,
+        chainId: network.chainId,
+        completion: completion
+      )
+    } else if token.isErc721 {
+      erc721TokenBalance(
+        token.contractAddress,
+        tokenId: token.tokenId,
+        accountAddress: accountAddress,
+        chainId: network.chainId,
+        completion: completion
+      )
+    } else {
+      let errorMessage = "Unable to fetch ethereum balance for \(token.symbol) token in account address '\(accountAddress)'"
```

Comment:
Currently, this error message will not see the user. This function is private to this file, and when status is not `success` we return nil value in the completion blocks on [L30](https://github.com/brave/brave-ios/pull/5726/files#diff-318bfcb7b1ee7544b84a0e90c73b12278955ea7b4e1f462f6ad5aa93c49ec33cR30) & [L129](https://github.com/brave/brave-ios/pull/5726/files#diff-318bfcb7b1ee7544b84a0e90c73b12278955ea7b4e1f462f6ad5aa93c49ec33cR129) of this file